core/src/main/java/org/infinispan/interceptors/SyncInvocationStage.java

```diff
@@ -36,6 +38,51 @@ public boolean isDone() {
       return CompletableFuture.completedFuture(rv);
    }
 
+   @Override
+   public Object thenApply(InvocationContext ctx, VisitableCommand command,
+                           InvocationSuccessFunction function) {
+      try {
+         return function.apply(ctx, command, rv);
+      } catch (Throwable throwable) {
+         return new SimpleAsyncInvocationStage(throwable);
```

Comment:
`InvocationStage` is the base class, it's not instantiable. I made `SyncInvocationStage` deal only with successful results to keep it simpler and easier to inline, but TBH I'm not sure how well that works nowadays.

Reply:
my idea is to have a "SyncInvocationStage" but for exceptions. 
The `AsyncInvocationStage` always instantiate a CompletableFuture and I bet there are cases where it isn't needed.

Reply:
I did something like this for the non-blocking xsite: https://github.com/infinispan/infinispan/pull/7276/files#diff-d963bfffb3f2513055256b1f8f3b84c1

Reply:
> my idea is to have a "SyncInvocationStage" but for exceptions.
> The `AsyncInvocationStage` always instantiate a CompletableFuture and I bet there are cases where it isn't needed.

IMO it's not worth it, because sync exceptions are not that common. The one exception that can happen a lot is `OutdatedTopologyException`, but it's almost always a result of an RPC so a `SyncExceptionStage` impl wouldn't help.

Reply:
I hit `Comment` before reading your PR... if you think it's common enough you can add an exceptional impl, but I remember when I initially wrote it I thought having fewer implementations would help the JIT do better inlining. I definitely don't think this is worth it when you can return a constant `SyncInvocationStage` instead:

```
     return rv == null ? NullSyncInvocationStage.getInstance() : new SyncInvocationStage(rv);
```

Reply:
@danberindei can you make your comments on https://github.com/infinispan/infinispan/pull/7276 ? 
I'll rebase it on top of yours